backend/ping.py

```diff
@@ -0,0 +1,178 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+###############################################################################
+
+
+'''`backend.ping`
+Efficiently scans given endpoints (pairs of hosts with ports) if they
+are connectible.
+
+The internal mechanism opens a large number of socket file descriptors
+with O_NONBLOCK flag set (see: socket(2), fcntl(2), connect(2)),
+and concurrently waits for each socket's state changes with pselect
+(see: select(2)).
+
+File descriptors creation is limited, so that the soft limit should
+not be exceeded (see: getrlimit(2)). If a large collection of endpoints
+is being scanned, it is split into smaller chunks. If the number
+of open file descriptors is dangerously high, scanning is being
+performed sequentially.
+'''
+
+
+import copy
+import errno
+import fcntl
+import itertools
+import os
+import resource
+import select
+import socket
+
+
+_MAGIC_NUMBER_DEFAULT_PORT = 22
+_MAGIC_NUMBER_SPLIT_LOWER_THRESHOLD = .25
+_MAGIC_NUMBER_SPLIT_UPPER_THRESHOLD = .75
+
+
+def scan(endpoints, default_port = _MAGIC_NUMBER_DEFAULT_PORT):
+    '''Checks if given endpoints are connectible
+
+    :return: a dictionary of (host, port): bool is open.
+    :param endpoints: a collection of tuples (host, port) or just host.
+    :param default_port: if port not set explicitly it will be defaulted
+            to this value.
+    '''
+    results = {}
+    filled_endpoints = _fill_endpoints(endpoints, default_port)
+    for e in _split(filled_endpoints):
+        r = _scan(e)
+        results.update(r)
+    return results
+
+
+def _get_number_of_open_file_descriptors():
+    return len(os.listdir('/proc/{0}/fd'.format(str(os.getpid()))))
+
+
+def _get_file_descriptor_resource_limit():
+    soft, hard = resource.getrlimit(resource.RLIMIT_NOFILE)
+    return soft
+
+
+def _create_socket_2(address_family, socket_type, protocol, address_tuple):
+    try:
+        sock = socket.socket(address_family, socket_type, protocol)
+    except socket.error:
+        return None, False
+    sock_fd = sock.fileno()
+    fd_flags = fcntl.fcntl(sock_fd, fcntl.F_GETFL)
+    fd_flags |= os.O_NONBLOCK
+    fcntl.fcntl(sock_fd, fcntl.F_SETFL, fd_flags)
+    try:
+        sock.connect(address_tuple)
+    except socket.error as e:
+        if e.errno == errno.EINPROGRESS:
+            return sock, False
+        else:
+            sock.close()
+            return None, False
+    else:
+        sock.close()
+        return None, True
+
+
+def _create_socket(struct_getaddrinfo):
+    return _create_socket_2(struct_getaddrinfo[0],
+                              struct_getaddrinfo[1],
+                              struct_getaddrinfo[2],
+                              struct_getaddrinfo[4])
+
+
+def _wait_for_any_change(socket_fds):
+    select_results = select.select(socket_fds,
```

Comment:
this call is blocking until at least one of the fd is ready. but this means that it might not return, right? its not what we want because there are other sockets that are already available (the ones that return is_open = True). 
maybe we should pass 0 as the timeout for select, which means it will just to a poll and not block.